optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/variable/listener/support/VariableListenerSupport.java

```diff
@@ -134,27 +137,30 @@ public void linkVariableListeners() {
     // ************************************************************************
 
     public void resetWorkingSolution() {
-        for (VariableListenerNotifiable notifiable : notifiableList) {
+        for (VariableListenerNotifiable<Solution_> notifiable : notifiableList) {
             VariableListener<Solution_, ?> variableListener = notifiable.getVariableListener();
             variableListener.resetWorkingSolution(scoreDirector);
         }
     }
 
     public void clearWorkingSolution() {
-        for (VariableListenerNotifiable notifiable : notifiableList) {
+        for (VariableListenerNotifiable<Solution_> notifiable : notifiableList) {
             VariableListener<Solution_, ?> variableListener = notifiable.getVariableListener();
             variableListener.close();
         }
     }
 
     public void beforeEntityAdded(EntityDescriptor<Solution_> entityDescriptor, Object entity) {
-        List<VariableListenerNotifiable> notifiableList = sourceEntityToNotifiableMap.get(entityDescriptor);
-        for (VariableListenerNotifiable notifiable : notifiableList) {
-            Collection<VariableListenerNotification> notificationQueue = notifiable.getNotificationQueue();
-            boolean added = notificationQueue.add(
-                    new VariableListenerNotification(entity, VariableListenerNotificationType.ENTITY_ADDED));
-            if (added) {
-                notifiable.getVariableListener().beforeEntityAdded(scoreDirector, entity);
+        List<VariableListenerNotifiable<Solution_>> notifiableList = sourceEntityToNotifiableMap.get(entityDescriptor);
+        if (!notifiableList.isEmpty()) {
+            VariableListenerNotification notification =
+                    new VariableListenerNotification(entity, VariableListenerNotificationType.ENTITY_ADDED);
+            for (VariableListenerNotifiable<Solution_> notifiable : notifiableList) {
+                Collection<VariableListenerNotification> notificationQueue = notifiable.getNotificationQueue();
+                boolean added = notificationQueue.add(notification);
+                if (added) {
+                    notifiable.getVariableListener().beforeEntityAdded(scoreDirector, entity);
+                }
             }
```

Comment:
This is the meat. The class is now created out of the loop and reused; it is immutable, it can be safely shared.